File: launcher/minimal/src/main/java/ink/reactor/launcher/logger/type/ConsoleLogger.java
package ink.reactor.launcher.logger.type;

import ink.reactor.launcher.logger.data.LoggerLevels;
import ink.reactor.launcher.logger.data.StyleLog;
import lombok.RequiredArgsConstructor;

import java.io.PrintWriter;

@RequiredArgsConstructor
public final class ConsoleLogger {
    private final PrintWriter writer;
    private final LoggerLevels loggerLevels;
    private final StyleLog debugStyle, logStyle, infoStyle, warnStyle, errorStyle;

    public void debug(final String prefix, final String message) {
        if (loggerLevels.debug()) {
            writer.println(debugStyle.style(prefix, message));
        }
    }

    public void log(final String prefix, final String message) {
        if (loggerLevels.log()) {
            writer.println(logStyle.style(prefix, message));
        }
    }

    public void info(final String prefix, final String message) {
        if (loggerLevels.info()) {
            writer.println(infoStyle.style(prefix, message));
        }
    }

    public void warn(final String prefix, final String message) {
        if (loggerLevels.warn()) {
            writer.println(warnStyle.style(prefix, message));
        }
    }

    public void error(final String prefix, final String message) {
        if (loggerLevels.error()) {
            writer.println(errorStyle.style(prefix, message));
        }
    }

    public void error(final String prefix, final String message, final Throwable throwable) {
        if (loggerLevels.error()) {
            writer.println(errorStyle.style(prefix, message));
            throwable.printStackTrace(System.err);
        }
    }
}

File: launcher/minimal/src/main/kotlin/ink/reactor/launcher/logger/type/ConsoleLogger.kt
package ink.reactor.launcher.logger.type

import ink.reactor.launcher.logger.ConsoleStyle
import ink.reactor.launcher.logger.LoggerLevels
import java.io.PrintWriter

class ConsoleLogger(
    val writer: PrintWriter,
    val loggerLevels: LoggerLevels,

    val debugStyle: ConsoleStyle,
    val logStyle: ConsoleStyle,
    val infoStyle: ConsoleStyle,
    val warnStyle: ConsoleStyle,
    val errorStyle: ConsoleStyle,
) {

    fun debug(prefix: String, message: String) {
        if (loggerLevels.debug) {
            print(debugStyle, prefix, message)
        }
    }

    fun log(prefix: String, message: String) {
        if (loggerLevels.log) {
            print(logStyle, prefix, message)
        }
    }

    fun info(prefix: String, message: String) {
        if (loggerLevels.info) {
            print(infoStyle, prefix, message)
        }
    }

    fun warn(prefix: String, message: String) {
        if (loggerLevels.warn) {
            print(warnStyle, prefix, message)
        }
    }

    fun error(prefix: String, message: String) {
        if (loggerLevels.error) {
            print(errorStyle, prefix, message)
        }
    }

    fun error(prefix: String, message: String, throwable: Throwable) {
        if (loggerLevels.error) {
            print(errorStyle, prefix, message)
            throwable.printStackTrace(System.err)
        }
    }

    private fun print(style: ConsoleStyle, prefix: String, message: String) =
        writer.println(style.prefix + prefix + style.text + message + style.afterText);
}
